contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -441,11 +468,69 @@ static void
 zfs_key_config_free(zfs_key_config_t *config)
 {
 	free(config->homes_prefix);
+	free(config->runstatedir);
+	free(config->homedir);
+	free(config->dsname);
+}
+
+static int
+find_dsname_by_prop_value(zfs_handle_t *zhp, void *data)
+{
+	zfs_key_config_t *target = data;
+	zfs_type_t type = zfs_get_type(zhp);
+	char mountpoint[ZFS_MAXPROPLEN];
+	char *dsname = NULL;
+
+	/* Iterate over any nested datasets */
+	if (zfs_iter_filesystems(zhp, find_dsname_by_prop_value, data) != 0) {
+		zfs_close(zhp);
+		return (1);
+	}
+
+	/* Skip any datasets whose type does not match */
+	if ((type & ZFS_TYPE_FILESYSTEM) == 0) {
+		zfs_close(zhp);
+		return (0);
+	}
+
+	dsname = strdup(zfs_get_name(zhp));
+	if (target->homes_prefix != NULL) {
+		if (strncmp(dsname, target->homes_prefix,
+		    strlen(target->homes_prefix)) != 0) {
+			free(dsname);
+			zfs_close(zhp);
+			return (0);
+		}
+	}
+
+	if (target->homedir != NULL) {
+		(void) zfs_prop_get(zhp, ZFS_PROP_MOUNTPOINT, mountpoint,
+		    sizeof (mountpoint), NULL, NULL, 0, B_FALSE);
+		if (strcmp(target->homedir, mountpoint) != 0) {
+			free(dsname);
+			zfs_close(zhp);
+			return (0);
+		}
+	}
+
+	target->dsname = dsname;
+	zfs_close(zhp);
+	return (1);
 }
 
 static char *
 zfs_key_config_get_dataset(zfs_key_config_t *config)
 {
+	if (config->homedir != NULL &&
+	    zfs_iter_root(g_zfs, find_dsname_by_prop_value, config)) {
+		if (config->dsname == NULL) {
+			return (NULL);
+		}
+		char *dsname = strdup(config->dsname);
```

Comment:
the `strdup` is wrong here. `config->dsname` is already malloced (by `strdup`)  in line 496. The present implementation has a memory leak caused by that. Removing the `strdup` also makes the special case for `NULL` unnecessary.

Reply:
I should have `strdup` removed from line 529, sorry for the churn.

One question for the case for `NULL` become unnecessary, if `config->dsname = NULL;` is removed from line 530 then `config->dsname` at some time loses its control to `strdup` from line 496. That means `config->dsname` might be pointing somewhere until `zfs_key_config_free()` is called?

What if `config->dsname = NULL;` on line 530 remain unchanged, and do `(config->dsname != NULL)? free(config->dsname):NULL;` in `zfs_key_config_free()`?

Reply:
According to the manpage of `free`:
> The free() function frees the memory space pointed to by ptr, which must have been returned by a previous call to malloc(), calloc(), or realloc().  Otherwise, or if free(ptr) has already been called before, undefined behavior
       occurs.  If ptr is NULL, no operation is performed.

Reading this documentation, `(config->dsname != NULL)? free(config->dsname):NULL;` has exactly the same effect as `free(config->dsname)`.

So we must ensure that pointers passed to `free` are either `NULL` or valid. So we must ensure that pointers in those fields are either `NULL` or valid. So my suggestion is to replace lines 526-531 with:
```
		char *dsname = config->dsname;
		config->dsname = NULL;
		return (dsname);
```

Reply:
Cool, will make the modification in next revision. Thanks.